# Assignment A1 - Data import

Course "Data processing and Visualization", IE500417, NTNU. Fall 2020

https://www.ntnu.edu/studies/courses/IE500417

**Note: plagiarism is strictly forbidden! You should never copy any source code from other students. If you use any code written by others (except the standards libraries: NumPy, SciPy, Pandas, etc), provide a reference.**

**If the teachers see that your work is mostly copy+paste from online code snippets, the grade can be reduced.**

**If a case of plagiarism is detected, it will be reported to the administration.**

## Task description

This assignment consists of several parts:
1. Set up development environment on your computer. It includes installation and setup of tools which will be necessary in other course assignments. 
2. Load data sets of several popular data formats: CSV, JSON, XML
3. Do simple checks and analysis of loaded data.

## Submission details

The assignment must be handed in on Blackboard. The following must be handed in:
1. Report in HTML or PDF format describing the results of this assignment. Preferably, it is generated from the Jupyter notebook you used (Hint: In Jupyter: File > Download as > HTML). Alternatively (if you use plain Python or other tools), prepare a readable report that contains figures and source code snippets necessary to understand your work.
2. Source code that you used to generate the results. This could be the the Jupyter notebook file, python source files, Matlab files, etc.

Deadlines and grading information on Blackboard.

---
## Detailed steps: Part 1
---

For this part you don't have to hand in anything. Without this part done you won't be able to do the rest of assignments.
The following tools should be set up:

### Anaconda 
Anaconda is a Python Data Science Toolset, including Pandas, NumPy and SciPy.
https://www.anaconda.com/products/individual#download-section .
 - Take Python 3.x version (not 2.x). Preferably 64-bit version (if your computer supports it)
 - Suggested to “Add Anaconda to System PATH” during installation
 - This will install Python 3.8 with all the packages 

### Jupyter notebook
It is an interactive web-based editor that allows you to write and test python code mixed with markup to format nice reports.
See http://jupyter.org/

You don’t have to install anything extra. Jupyter notebook is part of the Anaconda package.
Open the console (on Windows: open Anaconda Prompt) and run command

`jupyter notebook`

If you get the notebook started and a page opened in your browser, it is set up properly.

Jupyter considers one directory as its *root*.  is advised that you create one folder where you will put all your experimentation files and/or assignments. See here how to tell Jupyter to use your experiment folder as root: 
https://stackoverflow.com/questions/35254852/how-to-change-the-jupyter-start-up-folder 

To test that it is working, try opening the jupyter notebook (.ipynb file) provided with this assignment.

### WebStorm (Optional) 

WebStorm is a convenient editor (IDE) for JavaScript: https://www.jetbrains.com/webstorm/
Again, this tool is not necessary, strictly speaking. But if you plan that the resulting project may be a web-site using Javascript visualization libraries (D3, Google Charts, or others), WebStorm is a very handy editor for web page creation.

* Download it at https://www.jetbrains.com/webstorm/download/
* Apply for academic student license to all JetBrains tools (WebStorm, PyCharm, etc.):
https://www.jetbrains.com/shop/eform/students , use your @[stud.]ntnu.no address
* When installing, activate using JetBrains account

You can use the editor for any HTML + Javascript + CSS projects. In this course it may be useful when you work with the interactive visualizations.

### PyCharm (optional)
PyCharm is an editor for Python: https://www.jetbrains.com/pycharm/download/
It may be sufficient with Jupyter notebook. But it is good to have a proper Python editot ready, in case you need to explore any external code or debug your own.

* Download it at https://www.jetbrains.com/pycharm/download
* Take the Professional edition
* When installing, activate using JetBrains account

---
## Detailed steps: Part 2
---

In this part you will practice to load data from popular formats into a Pandas DataFrame structure. [Pandas](https://pandas.pydata.org/) is a Python library for data processing: . It provides convenient data structures:
* [DataFrame](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html) - represents a table of data. You can select, process, fitler, sort etc rows and columns easily.
* [Series](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.html) - represents one column of data.

In addition to loading data, some of the exercises here include basic data filtering and querying commands.

**You can use the Jupyter notebook template (.ipynb file, provided with this assignment) as a starting point, fill in your code in it and hand in the resulting notebook.**

Jupyter notebook is part of the Anaconda package. Open the console (on Windows: open Anaconda Prompt) and run command 

```jupyter notebook```

If you are not familiar with Jupyter, read [this tutorial](https://www.packtpub.com/books/content/getting-started-jupyter-notebook-part-1) to get started.

Copy the CSV, XML and other files attached to this assignment to the same folder where you .ipynb notebook is.

Use the same notebook for all the steps in this assignment.


### Part 2.1: Import data from CSV file

First you must import the Pandas library. It tells the Python interpreter, what functions are available in the Pandas. to make the writing shorter, one typically loads the pandas library with an alias pd. (The following block is python source code.)

In [1]:
# import all functions from Pandas library, they will be accessible through alias pd
import pandas as pd;

# for example the read_csv function will be accessible as pd.read_csv(...)

Now when the Pandas library is imported, load data from the CSV file into a Python DataFrame object called `houses`. Use the .CSV file attached to this assignment: real_estate.csv .
(Data source: https://support.spatialkey.com/spatialkey-sample-csv-data/ )
It contains data about sold real estate (apartment, houses, etc).

**Pay attention to the texts below. Check out all the blocks where you have to insert code!**

In [2]:
# Write code to load the CSV data into a Pandas Dataframe
houses = pd.read_csv("real_estate.csv")

Print the first 5 rows of the dataframe.

In [3]:
# Your code here
houses.head()

street        city    zip state  beds  baths  sq__ft  \
0      3526 HIGH ST  SACRAMENTO  95838    CA     2      1     836   
1       51 OMAHA CT  SACRAMENTO  95823    CA     3      1    1167   
2    2796 BRANCH ST  SACRAMENTO  95815    CA     2      1     796   
3  2805 JANETTE WAY  SACRAMENTO  95815    CA     2      1     852   
4   6001 MCMAHON DR  SACRAMENTO  95824    CA     2      1     797   

          type                     sale_date  price   latitude   longitude  
0  Residential  Wed May 21 00:00:00 EDT 2008  59222  38.631913 -121.434879  
1  Residential  Wed May 21 00:00:00 EDT 2008  68212  38.478902 -121.431028  
2  Residential  Wed May 21 00:00:00 EDT 2008  68880  38.618305 -121.443839  
3  Residential  Wed May 21 00:00:00 EDT 2008  69307  38.616835 -121.439146  
4  Residential  Wed May 21 00:00:00 EDT 2008  81900  38.519470 -121.435768

Print the last 5 rows of the dataframe.

In [4]:
# Your code here
houses.tail()

street             city    zip state  beds  baths  sq__ft  \
980   9169 GARLINGTON CT       SACRAMENTO  95829    CA     4      3    2280   
981      6932 RUSKUT WAY       SACRAMENTO  95823    CA     3      2    1477   
982    7933 DAFFODIL WAY   CITRUS HEIGHTS  95610    CA     3      2    1216   
983     8304 RED FOX WAY        ELK GROVE  95758    CA     4      2    1685   
984  3882 YELLOWSTONE LN  EL DORADO HILLS  95762    CA     3      2    1362   

            type                     sale_date   price   latitude   longitude  
980  Residential  Thu May 15 00:00:00 EDT 2008  232425  38.457679 -121.359620  
981  Residential  Thu May 15 00:00:00 EDT 2008  234000  38.499893 -121.458890  
982  Residential  Thu May 15 00:00:00 EDT 2008  235000  38.708824 -121.256803  
983  Residential  Thu May 15 00:00:00 EDT 2008  235301  38.417000 -121.397424  
984  Residential  Thu May 15 00:00:00 EDT 2008  235738  38.655245 -121.075915

### Part 2.2: Import data from JSON file

In a very similar fashion you can load data from a JSON file into a DataFrame. The nice thing with DataFrame - once you have loaded data into it (no matter what the source file format is), you can use the same processing methods.

Your task: load the data from file `colors.json` into a DataFrame called `colors`. 

In [5]:
# Your code here
colors = pd.read_json('colors.json')

Now we print contents of the loaded data:

In [6]:
colors

color value
0      red  #f00
1    green  #0f0
2     blue  #00f
3     cyan  #0ff
4  magenta  #f0f
5   yellow  #ff0
6    black  #000

### Part 2.3: Import data from SQL

Python Pandas allows you to read data from an SQL database into a DataFrame. See function 
[pandas.read_sql_query()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_sql_query.html). 

In this assignment we will use SQLite database, which is a very simple database format, storing everything in a single file. The pandas.read_sql_query() method is very flexible - it takes an SQL query and a connection to SQL interface for any database supporting it. It is not limited to SQLite, you can query MySQL, Oracle and other databases in the same manner. Only the creation of the connection will differ.

Hint: A good (but rather long) tutorial is [here](https://www.dataquest.io/blog/python-pandas-databases/). Search for text "read_sql_query" on the page, you will find a short source code example. 

Your task: load data from the attached file `chinook.sqlite3`. Use the following query to select first five customers, load them into a DataFrame, call it `first_customers`:
```SELECT firstname, lastname, address FROM customers LIMIT 5;```

In [7]:
# Your code here
import sqlite3

# Establish database connection
conn = sqlite3.connect('chinook.sqlite3')

first_customers = pd.read_sql_query('SELECT firstname, lastname, address FROM customers LIMIT 5;', conn)

Now we print the results:

In [8]:
first_customers

FirstName     LastName                          Address
0       Luís    Gonçalves  Av. Brigadeiro Faria Lima, 2170
1     Leonie       Köhler          Theodor-Heuss-Straße 34
2   François     Tremblay                1498 rue Bélanger
3      Bjørn       Hansen                 Ullevålsveien 14
4  František  Wichterlová                    Klanova 9/506

### Part 2.4: Import data from XML file

This part can be more difficult than loading a CSV. The reason: there is no standard XML reader in Pandas. You will have to implement it yourself. Hint: Google "Pandas DataFrame load XML".

**Your task**: load data from attached file `books.xml`, into variable called `books`.

Hint: you may want to [read the file content into a string variable](https://prefetch.net/blog/index.php/2012/01/30/reading-a-file-into-a-python-string/) first using open() and read() functions.

In [9]:
# Your code here
# The xml.etree.ElementTree module implements a simple and efficient API for parsing and creating XML data.
# Reference: https://intellipaat.com/community/42421/python-xml-file-to-pandas-dataframe
import xml.etree.ElementTree as etree

tree = etree.parse("books.xml")

root = tree.getroot()

columns = ["id", "author", "title", "genre", "price", "publish_date", "description"]

# Create an empty dataframe with labels
books = pd.DataFrame(columns = columns)

# Insert data into empty dataframe
for node in root: 

    id = node.attrib.get("id")
    author = node.find("author").text
    title = node.find("title").text
    genre = node.find("genre").text
    price = node.find("price").text
    publish_date = node.find("publish_date").text
    description = node.find("description").text

    books = books.append(pd.Series([id, author, title, genre, price, publish_date, description], index = columns), ignore_index = True)

books

id                author                                   title  \
0   bk101  Gambardella, Matthew                   XML Developer's Guide   
1   bk102            Ralls, Kim                           Midnight Rain   
2   bk103           Corets, Eva                         Maeve Ascendant   
3   bk104           Corets, Eva                         Oberon's Legacy   
4   bk105           Corets, Eva                      The Sundered Grail   
5   bk106      Randall, Cynthia                             Lover Birds   
6   bk107        Thurman, Paula                           Splish Splash   
7   bk108         Knorr, Stefan                         Creepy Crawlies   
8   bk109          Kress, Peter                            Paradox Lost   
9   bk110          O'Brien, Tim   Microsoft .NET: The Programming Bible   
10  bk111          O'Brien, Tim           MSXML3: A Comprehensive Guide   
11  bk112           Galos, Mike  Visual Studio 7: A Comprehensive Guide   

              genre  price publish_date  \
0          Computer  44.95   2000-10-01   
1           Fantasy   5.95   2000-12-16   
2           Fantasy   5.95   2000-11-17   
3           Fantasy   5.95   2001-03-10   
4           Fantasy   5.95   2001-09-10   
5           Romance   4.95   2000-09-02   
6           Romance   4.95   2000-11-02   
7            Horror   4.95   2000-12-06   
8   Science Fiction   6.95   2000-11-02   
9          Computer  36.95   2000-12-09   
10         Computer  36.95   2000-12-01   
11         Computer  49.95   2001-04-16   

                                          description  
0   An in-depth look at creating applications \n  ...  
1   A former architect battles corporate zombies, ...  
2   After the collapse of a nanotechnology \n     ...  
3   In post-apocalypse England, the mysterious \n ...  
4   The two daughters of Maeve, half-sisters, \n  ...  
5   When Carla meets Paul at an ornithology \n    ...  
6   A deep sea diver finds true love twenty \n    ...  
7   An anthology of horror stories about roaches,\...  
8   After an inadvertant trip through a Heisenberg...  
9   Microsoft's .NET initiative is explored in \n ...  
10  The Microsoft MSXML3 parser is covered in \n  ...  
11  Microsoft Visual Studio 7 is explored in depth...

## Part 3: Quick data inspection

In this part you do practice simple processing of the loaded data. Python and Pandas is a very rich toolset, here you get just a glimpse of it.

In this part, all tasks are related to the `house` data that you loaded from the CSV file in Part 2. If you are wondering about what methods to use, see [Hints](#Hints) section.

**Task for you**: Select number of rows in the data set, store it in variable `csv_num_rows`.

In [10]:
# Your code here
csv_num_rows = len(houses.index)
print("Number of rows in the CSV file: %i" % csv_num_rows)

Number of rows in the CSV file: 985


You can select one column of the DataFrame, it will be a Pandas Series object. For example, to select Zip code column:

In [11]:
zipcode = houses["zip"]

**Task for you**: Now select the column "price" into a separate variable called `prices`.

In [12]:
# Your code here
prices = houses["price"]

To validate that you are doing correctly, we can use some tests. We know that the 10th price in the file is 98937 (one can simply open the CSV file in a text editor to verify that). Indexing in python starts at zero. It means that prices[9] should be 98937. If you get error after executing the next block of code, something is wrong in your program. There are many possible errors. Therefore before testing the real value, we check if `prices` is a list containing at least 10 values. P.S. The assert statement means "check if this condition is true. If it isn't raise an error".

In [13]:
assert type(prices) == pd.core.series.Series
assert len(prices) >= 10 
assert prices[9] == 98937

Accessing specific rows and items in a DataFrame by providing an index or a range of indices is a topic in itself. We suggest to read a tutorial on it, for example, the [official Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/indexing.html).

**Task for you**: Calculate mean and standard deviation for column "price", store it in variables `price_mean` and `price_std` respectively.

In [14]:
# Your code here
price_mean = prices.mean()
price_std = prices.std()

Pandas DataFrame and Series classes has a handy method `describe()`. We can use it to check whether our calculated mean value is correct:

In [15]:
prices.describe()

count       985.000000
mean     234144.263959
std      138365.839085
min        1551.000000
25%      145000.000000
50%      213750.000000
75%      300000.000000
max      884790.000000
Name: price, dtype: float64

By doing "their homework" the teachers found out that the mean value should be 234144.263959. We make a test to see if your value is the same.

There can be different weird behaviour with number rounding. Therefore, one general hint: never compare floating point numbers on exact values. Instead, you can either round them and compare, or convert them to a string and then compare. In this case, we check if the mean price rounded to integer is 234144.

In [16]:
assert round(price_mean) == 234144

Pandas DataFrame has very rich selection interface. For example, you can easily select all rows, where price equals to 106852. Or all rows where price > 800000. The syntax is a bit weird, just get used to it. :) 
First we create a filter. It will be a list containing True and False values. The i-th value will be True, if the i–th row satisfies the condition. Then we apply that filter to the dataframe and select only the necessary rows: those having True value in the filter. 

In [17]:
# Create a filter - check which rows have price >= 800000
expensive_filter = (houses["price"] >= 800000)

In [18]:
# Now select only those rows in houses DataFrame which satisfiy the filter condition
expensive_houses = houses[expensive_filter]

Now let's print the table showing the expensive houses

In [19]:
expensive_houses

street             city    zip state  beds  baths  sq__ft  \
157          315 JUMEL CT  EL DORADO HILLS  95762    CA     6      5       0   
334      3935 EL MONTE DR           LOOMIS  95650    CA     4      4    1624   
863      2982 ABERDEEN LN  EL DORADO HILLS  95762    CA     4      3       0   
864  9401 BARREL RACER CT           WILTON  95693    CA     4      3    4400   

            type                     sale_date   price   latitude   longitude  
157  Residential  Wed May 21 00:00:00 EDT 2008  830000  38.669931 -121.059580  
334  Residential  Tue May 20 00:00:00 EDT 2008  839000  38.813337 -121.133348  
863  Residential  Fri May 16 00:00:00 EDT 2008  879000  38.706692 -121.058869  
864  Residential  Fri May 16 00:00:00 EDT 2008  884790  38.415298 -121.194858

You can combine the filter and row selection in a single line by putting the filter directly inside the dataframe [] selector:

In [20]:
expensive_houses = houses[houses["price"] >= 800000]

**Task for you**: Now that you know how to get the statistics (mean, std), how to filter rows and select one column of a DataFrame, do the following:
* Find maximum price in the whole data set
* Find the most expensive house (the one having maximum price)
* Find how many square feet it has and store it in variable `most_expensive_area`

In [21]:
# Your code here

# Maximum price
max_price = prices.max()

filter_most_expensive = (prices == max_price) 

# Most expensive house
house_most_expensive = houses[filter_most_expensive]

most_expensive_area = house_most_expensive["sq__ft"].item()

Now we test if the value is correct.
Hint: if your result of selection is one cell in Pandas Series, not a number, you can use the .item() method to get the first cell in the Series object.
For example, if you have `area` object of type series, use 

```most_expensive_area = area.item() ```

to get it as a numeric value.

In [22]:
assert most_expensive_area == 4400

You can also do manipulations with all the values in the whole column. Or do an operation on two columns. For example, you can calculate price per square feet and assign it to a new column in the DataFrame:

In [23]:
houses["sqft_price"] = houses["price"] / houses["sq__ft"]

**Task for you**: find the house having the cheapest square meter price. Assign it to variable `cheapest_house`. The resulting variable should be a DataFrame consisting of a single row. Hint: use the selection/filtering techniques described above.

In [24]:
# Your code here

# Prices per square feet of all houses
sqft_prices = houses["sqft_price"]

# Lowest price per square feet
lowest_sqft_price = sqft_prices.min()

filter_lowest_sqft_price = (sqft_prices == lowest_sqft_price)

cheapest_house = houses[filter_lowest_sqft_price]

To verify that you found the right house, you can run this code block.

In [25]:
assert type(cheapest_house) == pd.core.frame.DataFrame
assert len(cheapest_house) == 1
assert cheapest_house["price"].item() == 2000 
assert cheapest_house["sq__ft"].item() == 5822

We can use function `pandas.Series.unique()` to get a list of values that a particular series (DataFrame column) takes. Let's take a look what values does the "type" column take:

In [26]:
houses["type"].unique()

array(['Residential', 'Condo', 'Multi-Family', 'Unkown'], dtype=object)

**Task for you**: Find out how many Multi-family houses had less than 3 bathrooms. Store that number in variable `num_few_br_multif`. P.S. by "find out" we mean "write code that finds out". If you find the value by just looking at the file content, it does not count :).

In [27]:
# Your code here
filter_few_br_multif = (houses["type"] == "Multi-Family") & (houses["baths"] < 3)

# Multi-family houses that had less than 3 bathrooms 
few_br_multif = houses[filter_few_br_multif]

num_few_br_multif = len(few_br_multif.index)
num_few_br_multif

6

**Task for you**: Convert area from square feet to square meters, add it as a column named `area_m2` to the same DataFrame object `houses`.

In [28]:
# Your code here
houses["area_m2"] = houses["sq__ft"] / 10.764

**Task for you**: A bit more challenging exercise: select the top five cities by the total area of real estate sold. Show them order by the area, in descending order. I.e, the result should be a Pandas Series with five rows where the index contains city names and the value column contains total area sold in that city. Store result in variable `top_five_cities`.

Hint: If you do the calculations correctly, the top #3 city should be Antelope with total area sold ~ 5262.86 m2.

In [29]:
# Your code here

# Rank cities by the total area of real estate sold
cities_ranked = houses.groupby(["city"]).sum().sort_values(["area_m2"], ascending=False).head().reset_index()

top_five_cities = pd.Series(cities_ranked["area_m2"].values, cities_ranked["city"].values)
top_five_cities

SACRAMENTO        56863.619472
ELK GROVE         19579.617243
ANTELOPE           5262.820513
ROSEVILLE          4587.235229
CITRUS HEIGHTS     4255.295429
dtype: float64

## Reflection

Please reflect on the following questions:
1. How did the assignment go? Was it easy or hard?
2. How many hours did you spend on it?
3. What was the most time-consuming part?
4. If you need to do similar things later in your professional life, how can you improve? How can you do it more efficiently?
5. Was there something you would expect to learn that this exercise did not include?
6. Was there something that does not make sense?

**Your answers here**
1. Assignment went smoothly. It was not very hard.
2. Between two to three hours.
3. The final task.
4. Get familiarised with python coding and libraries.
5. No.
6. No.

## Hints

The following Pandas functions may be handy
* [read_csv in Pandas](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) may be useful. There are lots of possible parameters to the function. You probably need only the parameter specifying filename.
* [DataFrame.head()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.head.html) and [DataFrame.tail()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.tail.html).
* Series.mean(), .std() and .max() functions for statistics
* [DataFrame.groupby()](https://pandas.pydata.org/pandas-docs/stable/groupby.html) for grouping rows together; [.sort_values()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html) for sorting the rows by calues in specific columns.

A good [tutorial with a generic XML reading example](http://www.austintaylor.io/lxml/python/pandas/xml/dataframe/2016/07/08/convert-xml-to-pandas-dataframe/) is provided by Austin Taylor. 

It is quite common for programmers to not remember all the functions and all possible parameters. One skill that is important to master: ability to quickly find documentation and examples on the internet. Several tips:
1. Try typing your question in Google. For example, if you wonder how to read a CSV file in this context, you can try to Google ["python Pandas read CSV DataFrame"](https://www.google.com/?q=python%20Pandas%20read%20CSV%20Dataframe) and with high chance one of the top 5 results will give you the answer. Just "Read CSV" will not be enough. You should mention the context. For the assignments of this course the context typically is "Python Pandas". 
2. [Stackoverflow](https://stackoverflow.com/) is one of the best resources for programming questions. It has a lot of different questions and answers. When you have a question, there is a 90% chance that someone has had exactly that question before and someone else has answered him/her. But typically it is enough to type the question in Google. All roads (i.e., programming questions) lead to StackOverflow anyway :).
3. Check out the official documentation of the libraries that you are using. If you wonder about some DataFrame functions, find the DataFrame page in Pandas documentation. Hint: typically Google is smart enough to bring you to the documentation page. If you simply type ["Pandas DataFrame"](https://www.google.no/search?q=pandas+dataframe&oq=pandas+dataframe), the first result will typically be the documentation page. Official documentation is typically the page with the most detailed information on exactly the class or function that you need to explore. If you wonder about the possible function parameters, etc, see the docs.